### PR TITLE
[MI-2388] Enhancements

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
1. Update message in modal: When an already linked project is linked again 
2. Added error handling to project unlink modals.
3. Change Board > Boards in serviceType dropdown
4. Unsubscribe > Delete in help text of slash command
5. Option to delete subscrptions on unlinking project
image

### CODE REVIEW

File: webapp/src/containers/Rhs/projectDetails/header.tsx [hunk 1]
```diff
@@ -58,22 +63,43 @@ const Header = ({projectDetails, showAllSubscriptions, handlePagination, setShow
...
     const handleConfirmUnlinkProject = () => {
-        makeApiRequestWithCompletionStatus(pluginConstants.pluginApiServiceConfigs.unlinkProject.apiServiceName, projectDetails);
+        makeApiRequestWithCompletionStatus(pluginConstants.pluginApiServiceConfigs.unlinkProject.apiServiceName, {...projectDetails, deleteSubscriptions} as APIRequestPayload);
+    };
+
+    // Update the modal when project unlinking fails
+    const handleActionsAfterUnlinkingProjectFailed = (err: ApiErrorResponse) => {
+        const errorMessage = utils.getErrorMessage(true, 'ConfirmationModal', err);
+        if (errorMessage === pluginConstants.messages.error.adminAccessError) {
+            setConfirmationModalDescription(pluginConstants.messages.error.adminAccessErrorForUnlinking);
```
Comment (Kshitij-Katiyar) [inline webapp/src/containers/Rhs/projectDetails/header.tsx:73]: Is this correct that you are showing the error In the description of the modal. Is there a need to create an error state for this.

File: webapp/src/containers/modals/LinkModal/index.tsx [hunk 2]
```diff
@@ -63,6 +65,9 @@ const LinkModal = () => {
         serviceName: pluginConstants.pluginApiServiceConfigs.createLink.apiServiceName,
         payload: formFields as LinkPayload,
         handleSuccess: () => {
+            if ((data as CreateLinkResponse)?.message === 'This project is already linked.') {
```
Comment (Kshitij-Katiyar) [inline webapp/src/containers/modals/LinkModal/index.tsx:68]: Create constants for this errors as i think they are coming from the APIs and that can change in the future.

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/plugin/api.go [hunk 3]
```diff
@@ -220,6 +229,43 @@ func (p *Plugin) handleUnlinkProject(w http.ResponseWriter, r *http.Request) {
...
 
+func (p *Plugin) handleDeleteAllSubscriptions(mattermostUserID, projectID string) (int, error) {
+	subscriptionList, err := p.Store.GetAllSubscriptions(mattermostUserID)
+	if err != nil {
+		p.API.LogError(constants.FetchSubscriptionListError, "Error", err.Error())
+		return http.StatusInternalServerError, err
+	}
+
+	for _, subscription := range subscriptionList {
+		if subscription.ProjectID == projectID && subscription.MattermostUserID == mattermostUserID {
```
Comment (ayusht2810) [inline server/plugin/api.go:240]: if we are fetching all subscriptions of a user with the id `mattermostUserID`. Do we need to check the id here again?

File: server/plugin/api.go [hunk 4]
```diff
@@ -220,6 +229,43 @@ func (p *Plugin) handleUnlinkProject(w http.ResponseWriter, r *http.Request) {
...
+				MergeResult:                  subscription.MergeResult,
+				NotificationType:             subscription.NotificationType,
+				AreaPath:                     subscription.AreaPath,
+			}); deleteErr != nil {
+				p.API.LogError(constants.DeleteSubscriptionError, "Error", deleteErr.Error())
+				return http.StatusInternalServerError, deleteErr
+			}
+		}
+	}
+	return http.StatusOK, nil
```
Comment (ayusht2810) [inline server/plugin/api.go:266]:
```suggestion
	}

	return http.StatusOK, nil
```

File: server/plugin/api.go [hunk 5]
```diff
@@ -736,6 +782,19 @@ func returnStatusOK(w http.ResponseWriter) {
...
 
+func returnStatusWithMessage(w http.ResponseWriter, statusCode int, message string) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(statusCode)
+	m := map[string]string{"message": message}
+	response, err := json.Marshal(m)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+	if _, err := w.Write(response); err != nil {
```
Comment (ayusht2810) [inline server/plugin/api.go:793]:
```suggestion
	}
	
	if _, err := w.Write(response); err != nil {
```

File: server/plugin/command.go [hunk 6]
```diff
@@ -64,10 +64,10 @@ func (p *Plugin) getAutoCompleteData() *model.AutocompleteData {
 	link.AddTextArgument("URL of the project to be linked", "[projectURL]", "")
 	azureDevops.AddCommand(link)
 
-	subscription := model.NewAutocompleteData(constants.CommandSubscription, "", "Add/list/unsubscribe subscriptions")
+	subscription := model.NewAutocompleteData(constants.CommandSubscription, "", "Add/list/delete subscriptions")
 	subscriptionAdd := model.NewAutocompleteData(constants.CommandAdd, "", "Add a new subscription")
 	subscriptionView := model.NewAutocompleteData(constants.CommandList, "", "List subscriptions")
-	subscriptionUnsubscribe := model.NewAutocompleteData(constants.CommandDelete, "", "Unsubscribe a subscription")
+	subscriptionUnsubscribe := model.NewAutocompleteData(constants.CommandDelete, "", "Delete a subscription")
```
Comment (ayusht2810) [inline server/plugin/command.go:70]: shouldn't we change the name of `subscriptionUnsubscribe` to `subscriptionDelete`?
similarly for `subscriptionView`

File: webapp/src/containers/Rhs/projectDetails/header.tsx [hunk 7]
```diff
@@ -88,6 +114,21 @@ const Header = ({projectDetails, showAllSubscriptions, handlePagination, setShow
         }
     }, [filter.serviceType]);
 
+    const handleCheckboxChange = (e: ChangeEvent<HTMLInputElement>) => {
+        setDeleteSubscriptions(e.target.checked);
+    };
+
+    const deleteSubscriptionsCheckbox = (
+        <div>
```
Comment (ayusht2810) [inline webapp/src/containers/Rhs/projectDetails/header.tsx:122]:
```suggestion
        <>
```

File: webapp/src/containers/Rhs/projectDetails/header.tsx [hunk 8]
```diff
@@ -96,9 +137,12 @@ const Header = ({projectDetails, showAllSubscriptions, handlePagination, setShow
                 onConfirm={handleConfirmUnlinkProject}
                 isLoading={isUnlinkProjectLoading}
                 confirmBtnText='Unlink'
-                description={`Are you sure you want to unlink ${projectName}?`}
+                description={confirmationModalDescription}
                 title='Confirm Project Unlink'
-            />
+                showErrorPanel={unlinkConfirmationModalError}
+            >
+                {showDeleteSubscriptionsCheckbox ? deleteSubscriptionsCheckbox : <></>}
```
Comment (ayusht2810) [inline webapp/src/containers/Rhs/projectDetails/header.tsx:144]:
```suggestion
                {showDeleteSubscriptionsCheckbox && deleteSubscriptionsCheckbox}
```

Comment (ayusht2810) [reply webapp/src/containers/Rhs/projectDetails/header.tsx:144]: check if this works

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 9]
```diff
@@ -57,13 +64,49 @@ const ProjectList = () => {
...
         handleSuccess: handleActionsAfterUnlinkingProject,
+        handleError: handleActionsAfterUnlinkingProjectFailed,
     });
 
+    const handleCheckboxChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        setDeleteSubscriptions(e.target.checked);
+    };
+
+    const deleteSubscriptionsCheckbox = (
+        <div>
```
Comment (ayusht2810) [inline webapp/src/containers/Rhs/projectList/index.tsx:100]: same

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 10]
```diff
@@ -57,13 +64,49 @@ const ProjectList = () => {
...
+            <input
+                type='checkbox'
+                id='deleteSubscriptions'
+                onChange={handleCheckboxChange}
+            />
+            <label>{'Delete all your subscriptions associated with this project'}</label>
```
Comment (ayusht2810) [inline webapp/src/containers/Rhs/projectList/index.tsx:106]: we can create a constant for this, as it is used in another component as well

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 11]
```diff
@@ -76,11 +119,14 @@ const ProjectList = () => {
...
-                    isLoading={getApiState(pluginConstants.pluginApiServiceConfigs.unlinkProject.apiServiceName, projectToBeUnlinked).isLoading}
+                    isLoading={getApiState(pluginConstants.pluginApiServiceConfigs.unlinkProject.apiServiceName, {...projectToBeUnlinked, deleteSubscriptions} as APIRequestPayload).isLoading}
                     confirmBtnText='Unlink'
-                    description={`Are you sure you want to unlink ${projectToBeUnlinked?.projectName}?`}
+                    description={confirmationModalDescription}
                     title='Confirm Project Unlink'
-                />
+                    showErrorPanel={unlinkConfirmationModalError}
+                >
+                    {showDeleteSubscriptionsCheckbox ? deleteSubscriptionsCheckbox : <></>}
```
Comment (ayusht2810) [inline webapp/src/containers/Rhs/projectList/index.tsx:128]: same

File: webapp/src/containers/modals/LinkModal/index.tsx [hunk 12]
```diff
@@ -34,12 +34,14 @@ const LinkModal = () => {
     // State variables
     const {visibility, organization, project} = getLinkModalState(state);
     const [showResultPanel, setShowResultPanel] = useState(false);
+    const [resultPanelHeader, setResultPanelHeader] = useState('Project linked successfully.');
```
Comment (ayusht2810) [inline webapp/src/containers/modals/LinkModal/index.tsx:37]: constant

File: webapp/src/containers/Rhs/projectDetails/header.tsx (hunk 1 above)
Comment (raghavaggarwal2308) [reply webapp/src/containers/Rhs/projectDetails/header.tsx:73]: No we need to display the message in the description. This will be displayed when a non admin user trie to delete a subscription. In such case we want the user to know that u cannot delete the subscription but can unlink the project

File: webapp/src/containers/Rhs/projectDetails/header.tsx (hunk 8 above)
Comment (raghavaggarwal2308) [reply webapp/src/containers/Rhs/projectDetails/header.tsx:144]: It gives type error that children cannot be boolean

File: docs/oauth_setup.md [hunk 13]
```diff
@@ -2,7 +2,7 @@
 
 ## Register your OAuth application on Azure DevOps
   - Log in to [AzureDevops](https://app.vsaex.visualstudio.com).
-  - Click on the "Create new application" button in the bottom left corner.
+  - Register your app by opening [this](https://app.vsaex.visualstudio.com/app/register) link.
```
Comment (avas27JTG) [inline docs/oauth_setup.md:5]:
```suggestion
  - Register your app by visiting [here](https://app.vsaex.visualstudio.com/app/register) link.
```

File: server/plugin/api.go [hunk 14]
```diff
@@ -220,6 +229,44 @@ func (p *Plugin) handleUnlinkProject(w http.ResponseWriter, r *http.Request) {
...
+				PullRequestCreatedBy:         subscription.PullRequestCreatedBy,
+				PullRequestReviewersContains: subscription.PullRequestReviewersContains,
+				PushedBy:                     subscription.PushedBy,
+				MergeResult:                  subscription.MergeResult,
+				NotificationType:             subscription.NotificationType,
+				AreaPath:                     subscription.AreaPath,
+			}); deleteErr != nil {
+				p.API.LogError(constants.DeleteSubscriptionError, "Error", deleteErr.Error())
+				return http.StatusInternalServerError, deleteErr
+			}
```
Comment (avas27JTG) [inline server/plugin/api.go:263]: make this a reusable function to be used both here and in delete a single subscription API

File: webapp/src/containers/Rhs/projectDetails/header.tsx [hunk 15]
```diff
@@ -88,6 +114,21 @@ const Header = ({projectDetails, showAllSubscriptions, handlePagination, setShow
...
+    };
+
+    const deleteSubscriptionsCheckbox = (
+        <>
+            <input
+                type='checkbox'
+                id='deleteSubscriptions'
+                onChange={handleCheckboxChange}
+            />
+            <label>{pluginConstants.common.deleteAllSubscriptionsMessage}</label>
```
Comment (avas27JTG) [inline webapp/src/containers/Rhs/projectDetails/header.tsx:128]: wrap these into a `div` with `flex` and vertically align them center, also put a gap of `5px` between checkbox and lebel

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 16]
```diff
@@ -57,13 +64,49 @@ const ProjectList = () => {
...
         serviceName: pluginConstants.pluginApiServiceConfigs.unlinkProject.apiServiceName,
-        payload: projectToBeUnlinked,
+        payload: {...projectToBeUnlinked, deleteSubscriptions} as APIRequestPayload,
         handleSuccess: handleActionsAfterUnlinkingProject,
+        handleError: handleActionsAfterUnlinkingProjectFailed,
     });
 
+    const handleCheckboxChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        setDeleteSubscriptions(e.target.checked);
+    };
```
Comment (avas27JTG) [inline webapp/src/containers/Rhs/projectList/index.tsx:97]:
```suggestion
    const handleCheckboxChange = (e: React.ChangeEvent<HTMLInputElement>) => setDeleteSubscriptions(e.target.checked);
```

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 17]
```diff
@@ -57,13 +64,49 @@ const ProjectList = () => {
...
+
+    const deleteSubscriptionsCheckbox = (
+        <>
+            <input
+                type='checkbox'
+                id='deleteSubscriptions'
+                onChange={handleCheckboxChange}
+            />
+            <label>{pluginConstants.common.deleteAllSubscriptionsMessage}</label>
+        </>
```
Comment (avas27JTG) [inline webapp/src/containers/Rhs/projectList/index.tsx:107]: same

File: webapp/src/pluginConstants/common.ts [hunk 18]
```diff
@@ -5,6 +5,11 @@ export const RightSidebarHeader = 'Azure DevOps';
 
 export const MMCSRF = 'MMCSRF';
 export const HeaderCSRFToken = 'X-CSRF-Token';
+export const StatusCodeForbidden = 403;
+
+export const deleteAllSubscriptionsMessage = 'Delete all your subscriptions associated with this project';
```
Comment (avas27JTG) [inline webapp/src/pluginConstants/common.ts:10]:
```suggestion
export const deleteAllSubscriptionsMessage = 'Delete all your subscriptions created for this project';
```

File: webapp/src/pluginConstants/messages.ts [hunk 19]
```diff
@@ -13,4 +13,6 @@ export const error = {
 
     // Link
     notAccessibleError: 'Looks like this project/organization does not exist or you do not have permissions to access it',
+    adminAccessErrorForUnlinking: 'You do not have the permissions to delete subscriptions but you can still unlink the project',
```
Comment (avas27JTG) [inline webapp/src/pluginConstants/messages.ts:16]:
```suggestion
    adminAccessErrorForUnlinking: 'You do not have sufficient permissions to delete subscriptions for this project but you can still unlink the project',
```

File: server/plugin/api.go [hunk 20]
```diff
@@ -690,32 +673,39 @@ func (p *Plugin) handleDeleteSubscriptions(w http.ResponseWriter, r *http.Reques
...
-		ChannelID:                    body.ChannelID,
-		Repository:                   body.Repository,
-		TargetBranch:                 body.TargetBranch,
-		PullRequestCreatedBy:         body.PullRequestCreatedBy,
-		PullRequestReviewersContains: body.PullRequestReviewersContains,
-		PushedBy:                     body.PushedBy,
-		MergeResult:                  body.MergeResult,
-		NotificationType:             body.NotificationType,
-		AreaPath:                     body.AreaPath,
+		MattermostUserID:             subscription.MattermostUserID,
```
Comment (ayusht2810) [inline server/plugin/api.go:691]: can't we directly send `subscription` here from the arguments?

File: server/plugin/api.go [hunk 21]
```diff
@@ -690,32 +673,39 @@ func (p *Plugin) handleDeleteSubscriptions(w http.ResponseWriter, r *http.Reques
...
+		NotificationType:             subscription.NotificationType,
+		AreaPath:                     subscription.AreaPath,
 	}); deleteErr != nil {
-		p.API.LogError(constants.DeleteSubscriptionError, "Error", deleteErr.Error())
-		p.handleError(w, r, &serializers.Error{Code: http.StatusInternalServerError, Message: deleteErr.Error()})
+		return http.StatusInternalServerError, deleteErr
 	}
 
-	returnStatusOK(w)
+	return http.StatusOK, nil
```
Comment (ayusht2810) [inline server/plugin/api.go:708]: why this change?

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api.go (hunk 21 above)
Comment (raghavaggarwal2308) [reply server/plugin/api.go:695]: This is not a change we create a separate function. Github got little confused here

File: server/plugin/api.go [hunk 22]
```diff
@@ -691,32 +673,39 @@ func (p *Plugin) handleDeleteSubscriptions(w http.ResponseWriter, r *http.Reques
...
 	}
 
-	if statusCode, err := p.Client.DeleteSubscription(body.Organization, subscription.SubscriptionID, mattermostUserID); err != nil {
-		p.API.LogError(constants.DeleteSubscriptionError, "Error", err.Error())
-		p.handleError(w, r, &serializers.Error{Code: statusCode, Message: err.Error()})
-		return
+	statusCode, deleteErr := p.deleteSubscription(subscription, mattermostUserID)
+	if deleteErr != nil {
+		p.API.LogError(constants.DeleteSubscriptionError, "Error", deleteErr.Error())
+		p.handleError(w, r, &serializers.Error{Code: statusCode, Message: deleteErr.Error()})
```
Comment (avas27JTG) [inline server/plugin/api.go:679]: should `return` after error from here

**Review by avas27JTG — APPROVED** [review]

**Review by avas27JTG — APPROVED** [review]

**Review by ayusht2810 — APPROVED** [review]
